text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+
+Creating a gem content ingestion and storage system will enable rubygems.org to support the security and useability gems in the Ruby Gems ecosystem.
+The gem content repository will be the authoritative source for the content of gems, accessible as individual files without without requiring downloading and unpacking a gem.
+
+In the near future, the indexed gem content will allow an API for accessing the gem content.
+The API will support both rubygems.org and external projects that wish to use individual gem content.
+Rubygems.org plans to add tools for browsing gem content on rubygems.org or through the `gem` command line interface.
+
+# Guide-level explanation
+
+For the purpose of illustration, I will use the gem `rake` as an example.
+
+When a gem is uploaded, rubygems.org saves it as a tarball file named `rake-13.0.6.gem`.
+A gem may be installed with `gem install` or the `.gem` file can be downloaded using `gem fetch`.
+Downloading the entire gem locally is currently the only way to access the contents of gem.
+
+With the gem content ingestion system in place, gem contents will be saved to S3 and indexed in the rubygems.org database.
+Binary files will be recorded in metadata, but not stored.
+Gem contents will be stored at an S3 key named for the SHA256 of the file to de-duplicate files within a gem.
+Yanking will remove file contents unique to the gem, as expected, without altering unyanked gem contents.
+
+### Background process
+
+In order to ingest the contents of a gem, a background process can be enqueued to index and store the gem contents.
+
+At gem upload time, or on demand, a background worker will download, unpack, record file metadata, index files in the database as rows, and upload each file to a content-only S3 bucket.
+The process will result in a row in the database for each file in a gem, with the relative path, checksum, and size of every file.
+
+### Indexing files to create a manifest
+
+The background worker will create a json document manifest of all the files and store it in S3 alongside the files themselves.
+
+The manifest will hold the path, checksum, size, content type, binary determination, and plain text line count.
+
+In order to support yanking gems, which requires comparing checksums across all versions of a gem, a table will be added to the rubygems.org database.
+The table will consist of one row per gem version with a foreign key for the `version_id` and a jsonb column holding only the SHA256 checksums of all files in that version.
+
+The table will be indexed on `version_id` to support looking up all checksums in all versions of a gem.
+This table is for the sole purpose of yanking a gem, discussed below.
+
+### Recording, but not storing binary files
+
+Files that are identified as binary will be flagged in the index and not uploaded to the storage bucket.
```

Comment:
how will we identify binary files?

Reply:
The idea I like best is using grep, since the algorithm for determining which files are "binary" will match with most developers experience of files for which grep doesn't return results.